controllers/workspace_controller_variable_sets.go

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	tfc "github.com/hashicorp/go-tfe"
+)
+
+func (w *workspaceInstance) getVariableSetIDs(ctx context.Context) ([]string, error) {
+	variableSets := w.instance.Spec.VariableSets
+
+	if variableSets == nil {
+		return nil, fmt.Errorf("'spec.VariableSets' is not set")
+	}
+
+	var variableSetIDs []string
+	var namesToLookup []string
+
+	// Fetch all IDs and names to look up
+	for _, vs := range variableSets {
+		if vs.ID != "" {
+			variableSetIDs = append(variableSetIDs, vs.ID)
+		}
+		if vs.Name != "" {
+			namesToLookup = append(namesToLookup, vs.Name)
+		}
+	}
+
+	if len(namesToLookup) > 0 {
+		// Get all variable sets to find those referred by name
+		listOpts := &tfc.VariableSetListOptions{
+			ListOptions: tfc.ListOptions{
+				PageSize: maxPageSize,
+			},
+		}
+
+		// Get all var sets at once to optimize soln, a single API call
+		for {
+			variableSetList, err := w.tfClient.Client.VariableSets.List(ctx, w.instance.Spec.Organization, listOpts)
+			if err != nil {
+				return nil, err
+			}
+
+			// Searching by name instead
+			for _, vs := range variableSetList.Items {
+				for _, name := range namesToLookup {
+					if vs.Name == name {
+						variableSetIDs = append(variableSetIDs, vs.ID)
+					}
+				}
+			}
+
+			// Break if there are no more pages to fetch
+			if variableSetList.NextPage == 0 {
+				break
+			}
+
+			// Update page number to fetch the next set of results
+			listOpts.PageNumber = variableSetList.NextPage
+		}
+	}
+
+	if len(variableSetIDs) == 0 {
+		return nil, fmt.Errorf("no valid Variable Sets found in 'spec.VariableSets'")
+	}
+
+	return variableSetIDs, nil
+}
+
+func (r *WorkspaceReconciler) reconcileVariableSets(ctx context.Context, w *workspaceInstance, workspace *tfc.Workspace) error {
+	// Get the ID for the variable sets specified in the spec
+	variableSetIDs, err := w.getVariableSetIDs(ctx)
+	if err != nil {
+		return err
+	}
+
+	if len(variableSetIDs) == 0 {
+		w.log.Info("Reconcile Variable Set", "msg", "no variable sets to reconcile")
+		return nil
+	}
+
+	// List all variable sets once, a single API call
+	listOpts := &tfc.VariableSetListOptions{
```

Comment:
Page size = 100, what happens if there are >100?
for loop to iterate thru all pages to ensure we access everything...